Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Beckn/ACL/Update.hs

```diff
@@ -14,58 +14,134 @@
+import qualified Kernel.Types.Beckn.Gps as Gps
 import Kernel.Types.Id
 import qualified Kernel.Types.Registry.Subscriber as Subscriber
 import Kernel.Utils.Common
 import Tools.Error (GenericError (InvalidRequest))
 
 buildUpdateReq ::
-  ( HasFlowEnv m r '["coreVersion" ::: Text]
+  ( HasFlowEnv m r '["_version" ::: Text]
   ) =>
   Subscriber.Subscriber ->
-  Update.UpdateReq ->
+  Spec.UpdateReq ->
   m DUpdate.DUpdateReq
 buildUpdateReq subscriber req = do
-  validateContext Context.UPDATE $ req.context
-  unless (subscriber.subscriber_id == req.context.bap_id) $
+  ContextV2.validateContext Context.UPDATE $ req.updateReqContext
+  bap_uri <- Common.getContextBapUri req.updateReqContext
+  unless (Just subscriber.subscriber_id == req.updateReqContext.contextBapId) $
     throwError (InvalidRequest "Invalid bap_id")
-  unless (subscriber.subscriber_url == req.context.bap_uri) $
+  unless (subscriber.subscriber_url == bap_uri) $
     throwError (InvalidRequest "Invalid bap_uri")
-  pure $ parseEvent req.message.order
+  parseEventV2 req.updateReqMessage
 
-parseEvent :: Update.UpdateEvent -> DUpdate.DUpdateReq
-parseEvent (Update.PaymentCompleted pcEvent) = do
-  DUpdate.PaymentCompletedReq
-    { bookingId = Id pcEvent.id,
-      rideId = Id pcEvent.fulfillment.id,
-      paymentStatus = castPaymentStatus pcEvent.payment.status,
-      paymentMethodInfo = mkPaymentMethodInfo pcEvent.payment
-    }
-parseEvent (Update.EditLocation elEvent) = do
-  DUpdate.EditLocationReq
-    { bookingId = Id elEvent.id,
-      rideId = Id elEvent.fulfillment.id,
-      origin = elEvent.fulfillment.origin.location,
-      destination = elEvent.fulfillment.destination.location
-    }
+parseEventV2 :: (MonadFlow m) => Spec.UpdateReqMessage -> m DUpdate.DUpdateReq
+parseEventV2 req = do
+  eventType <-
+    req.updateReqMessageOrder.orderFulfillments
+      >>= listToMaybe
+      >>= (.fulfillmentState)
+      >>= (.fulfillmentStateDescriptor)
+      >>= (.descriptorCode)
+      & fromMaybeM (InvalidRequest "Event type is not present in OnUpdateReq.")
+  case eventType of
+    "PAYMENT_COMPLETED" -> parsePaymentCompletedEvent req
+    "EDIT_LOCATION" -> parseEditLocationEvent req
+    _ -> throwError (InvalidRequest "Invalid event type")
 
-mkPaymentMethodInfo :: Update.Payment -> DMPM.PaymentMethodInfo
-mkPaymentMethodInfo Update.Payment {..} =
-  DMPM.PaymentMethodInfo
-    { collectedBy = Common.castPaymentCollector collected_by,
-      paymentType = Common.castPaymentType _type,
-      paymentInstrument = Common.castPaymentInstrument instrument
-    }
+parsePaymentCompletedEvent :: (MonadFlow m) => Spec.UpdateReqMessage -> m DUpdate.DUpdateReq
+parsePaymentCompletedEvent req = do
+  rideId <- req.updateReqMessageOrder.orderFulfillments >>= listToMaybe >>= (.fulfillmentId) & fromMaybeM (InvalidRequest "RideId not present")
+  headPayment <- req.updateReqMessageOrder.orderPayments >>= listToMaybe & fromMaybeM (InvalidRequest "Payment not present")
```

Comment:
why `headPayment` and not simply `payment` since only 1 payment info will be present